### PR TITLE
[Feature] Add task_modules type hint

### PR DESCRIPTION
**Pull request by A-new-b**
## Motivation

Add task_modules type hint to the code and make it more readable.

## Modification

add type hints to mmdet3d/model/task_modules

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
Thanks for your PR. Please review this PR. @Xiangxu-0103

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 1]
```diff
@@ -153,12 +162,12 @@ def single_level_grid_anchors(self, featmap_size, scale, device='cuda'):
...
-                             anchor_range,
-                             scale=1,
-                             sizes=[[3.9, 1.6, 1.56]],
+                             feature_size: Union[List[float], Tuple[float]],
+                             anchor_range: Union[Tensor, List[float]],
+                             scale: Union[float, int] = 1,
+                             sizes: Union[Tensor, int] = [[3.9, 1.6, 1.56]],
                              rotations=[0, 1.5707963],
-                             device='cuda'):
+                             device='cuda') -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:170]:
```suggestion
                             device: Union[str, torch.device] = 'cuda') -> Tensor:
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 2]
```diff
@@ -75,18 +79,20 @@ def __repr__(self):
...
     @property
-    def num_levels(self):
+    def num_levels(self) -> int:
         """int: Number of feature levels that the generator is applied to."""
         return len(self.scales)
 
-    def grid_anchors(self, featmap_sizes, device='cuda'):
+    def grid_anchors(self,
+                     featmap_sizes: List[tuple],
+                     device: str = 'cuda') -> List[Tensor]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:95]:
```suggestion
                     device: Union[str, torch.device] = 'cuda') -> List[Tensor]:
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 3]
```diff
@@ -112,7 +118,10 @@ def grid_anchors(self, featmap_sizes, device='cuda'):
             multi_level_anchors.append(anchors)
         return multi_level_anchors
 
-    def single_level_grid_anchors(self, featmap_size, scale, device='cuda'):
+    def single_level_grid_anchors(self,
+                                  featmap_size: Tuple[int],
+                                  scale: float,
+                                  device: str = 'cuda') -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:124]:
```suggestion
                                  device: Union[str, torch.device] = 'cuda') -> Tensor:
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 4]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
         self.align_corner = align_corner
 
     def anchors_single_range(self,
-                             feature_size,
-                             anchor_range,
-                             scale,
-                             sizes=[[3.9, 1.6, 1.56]],
-                             rotations=[0, 1.5707963],
-                             device='cuda'):
+                             feature_size: Union[List[float], Tuple[float]],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:265]:
```suggestion
                             feature_size: List[int],
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 5]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
+                             anchor_range: Union[Tensor, List[float]],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:266]:
```suggestion
                             anchor_range: List[float],
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 6]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
+                             scale: Union[float, int],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:267]:
```suggestion
                             scale: int,
```

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 7]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
+                             sizes: Union[List[list], np.ndarray,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:268]: sizes: Union[List[List[float]], List[float]]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 8]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
+                                          Tensor] = [[3.9, 1.6, 1.56]],
+                             rotations: Union[List[float], np.ndarray,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:270]: rotations: List[float]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 9]
```diff
@@ -248,17 +257,19 @@ class AlignedAnchor3DRangeGenerator(Anchor3DRangeGenerator):
...
+                                              Tensor] = [0, 1.5707963],
+                             device: str = 'cuda') -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:272]: device: Union[str, torch.device]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 10]
```diff
@@ -352,12 +363,14 @@ class AlignedAnchor3DRangeGeneratorPerCls(AlignedAnchor3DRangeGenerator):
...
 
-    def __init__(self, **kwargs):
+    def __init__(self, **kwargs) -> None:
         super(AlignedAnchor3DRangeGeneratorPerCls, self).__init__(**kwargs)
         assert len(self.scales) == 1, 'Multi-scale feature map levels are' + \
             ' not supported currently in this kind of anchor generator.'
 
-    def grid_anchors(self, featmap_sizes, device='cuda'):
+    def grid_anchors(self,
+                     featmap_sizes: List[tuple],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:372]: List[Tuple[int]]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 11]
```diff
@@ -352,12 +363,14 @@ class AlignedAnchor3DRangeGeneratorPerCls(AlignedAnchor3DRangeGenerator):
...
+                     device: str = 'cuda') -> List[List[Tensor]]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:373]: Union[str, torch.device]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 12]
```diff
@@ -379,7 +392,10 @@ def grid_anchors(self, featmap_sizes, device='cuda'):
         multi_level_anchors.append(anchors)
         return multi_level_anchors
 
-    def multi_cls_grid_anchors(self, featmap_sizes, scale, device='cuda'):
+    def multi_cls_grid_anchors(self,
+                               featmap_sizes: List[tuple],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:396]: List[Tuple[int]]

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 13]
```diff
@@ -379,7 +392,10 @@ def grid_anchors(self, featmap_sizes, device='cuda'):
...
+                               scale: float,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:397]: int

File: mmdet3d/models/task_modules/anchor/anchor_3d_generator.py [hunk 14]
```diff
@@ -379,7 +392,10 @@ def grid_anchors(self, featmap_sizes, device='cuda'):
...
+                               device: str = 'cuda') -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/anchor/anchor_3d_generator.py:398]:
```suggestion
                               device: Union[str, torch.device] = 'cuda') -> List[Tensor]:
```

File: mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py [hunk 15]
```diff
@@ -22,13 +25,13 @@ class CenterPointBBoxCoder(BaseBBoxCoder):
...
-                 out_size_factor,
-                 voxel_size,
-                 post_center_range=None,
-                 max_num=100,
-                 score_threshold=None,
-                 code_size=9):
+                 pc_range: List[float],
+                 out_size_factor: int,
+                 voxel_size: List[float],
+                 post_center_range: List[float] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py:31]: Optional[List[float]]

File: mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py [hunk 16]
```diff
@@ -22,13 +25,13 @@ class CenterPointBBoxCoder(BaseBBoxCoder):
...
+                 max_num: int = 100,
+                 score_threshold: float = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py:33]: Optional[float]

File: mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py [hunk 17]
```diff
@@ -38,7 +41,10 @@ def __init__(self,
         self.score_threshold = score_threshold
         self.code_size = code_size
 
-    def _gather_feat(self, feats, inds, feat_masks=None):
+    def _gather_feat(self,
+                     feats: Tensor,
+                     inds: Tensor,
+                     feat_masks: Tensor = None) -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py:47]: feat_masks: Optional[Tensor]

File: mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py [hunk 18]
```diff
@@ -115,14 +121,14 @@ def encode(self):
...
-               reg=None,
-               task_id=-1):
+               heat: Tensor,
+               rot_sine: Tensor,
+               rot_cosine: Tensor,
+               hei: Tensor,
+               dim: Tensor,
+               vel: Tensor,
+               reg: Tensor = None,
+               task_id: int = -1) -> List[Dict]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/centerpoint_bbox_coders.py:131]: List[Dict[str, Tensor]]

File: mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py [hunk 19]
```diff
@@ -22,10 +25,10 @@ class FCOS3DBBoxCoder(BaseBBoxCoder):
     """
 
     def __init__(self,
-                 base_depths=None,
-                 base_dims=None,
-                 code_size=7,
-                 norm_on_bbox=True):
+                 base_depths: Tuple[Tuple[float]] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py:28]: Optional[Tuple[Tuple[float]]]

File: mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py [hunk 20]
```diff
@@ -22,10 +25,10 @@ class FCOS3DBBoxCoder(BaseBBoxCoder):
...
+                 base_dims: Tuple[Tuple[float]] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py:29]: Optional[Tuple[Tuple[float]]]

File: mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py [hunk 21]
```diff
@@ -36,7 +39,12 @@ def encode(self, gt_bboxes_3d, gt_labels_3d, gt_bboxes, gt_labels):
         # TODO: refactor the encoder in the FCOS3D and PGD head
         pass
 
-    def decode(self, bbox, scale, stride, training, cls_score=None):
+    def decode(self,
+               bbox: Tensor,
+               scale: tuple,
+               stride: int,
+               training: bool,
+               cls_score: Tensor = None) -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/fcos3d_bbox_coder.py:47]: cls_score: Optional[Tensor] = None

File: mmdet3d/models/task_modules/coders/groupfree3d_bbox_coder.py [hunk 22]
```diff
@@ -116,7 +121,11 @@ def decode(self, bbox_out, prefix=''):
         bbox3d = torch.cat([center, bbox_size, dir_angle], dim=-1)
         return bbox3d
 
-    def split_pred(self, cls_preds, reg_preds, base_xyz, prefix=''):
+    def split_pred(self,
+                   cls_preds: Tensor,
+                   reg_preds: Tensor,
+                   base_xyz: Tensor,
+                   prefix: str = ''):
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/groupfree3d_bbox_coder.py:128]:
```suggestion
                   prefix: str = '') -> Dict[str, Tensor]:
```

File: mmdet3d/models/task_modules/coders/monoflex_bbox_coder.py [hunk 23]
```diff
@@ -283,13 +288,15 @@ def decode_location(self,
...
-                        dimensions,
-                        cam2imgs,
-                        downsample_ratio=4,
-                        group0_index=[(7, 3), (0, 4)],
-                        group1_index=[(2, 6), (1, 5)]):
+    def keypoints2depth(
+            self,
+            keypoints2d: Tensor,
+            dimensions: Tensor,
+            cam2imgs: int,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/monoflex_bbox_coder.py:295]: Tensor

File: mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py [hunk 24]
```diff
@@ -18,7 +22,10 @@ class PointXYZWHLRBBoxCoder(BaseBBoxCoder):
             each class. Defaults to None.
     """
 
-    def __init__(self, code_size=7, use_mean_size=True, mean_size=None):
+    def __init__(self,
+                 code_size: int = 7,
+                 use_mean_size: bool = True,
+                 mean_size: List[List[float]] = None):
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py:28]: Optional[List[List]]

File: mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py [hunk 25]
```diff
@@ -28,7 +35,10 @@ def __init__(self, code_size=7, use_mean_size=True, mean_size=None):
                 f'The min of mean_size should > 0, however currently it is '\
                 f'{self.mean_size.min()}, please check it in your config.'
 
-    def encode(self, gt_bboxes_3d, points, gt_labels_3d=None):
+    def encode(self,
+               gt_bboxes_3d: BaseInstance3DBoxes,
+               points: Tensor,
+               gt_labels_3d: Tensor = None) -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py:41]: gt_labels_3d: Optional[Tensor] = None

File: mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py [hunk 26]
```diff
@@ -75,7 +85,10 @@ def encode(self, gt_bboxes_3d, points, gt_labels_3d=None):
              torch.sin(rg), *cgs],
             dim=-1)
 
-    def decode(self, box_encodings, points, pred_labels_3d=None):
+    def decode(self,
+               box_encodings: Tensor,
+               points: Tensor,
+               pred_labels_3d: Tensor = None) -> Tensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/point_xyzwhlr_bbox_coder.py:91]: pred_labels_3d: Optional[Tensor]

File: mmdet3d/models/task_modules/coders/smoke_bbox_coder.py [hunk 27]
```diff
@@ -50,12 +57,12 @@ def encode(self, locations, dimensions, orientations, input_metas):
...
-               labels,
-               cam2imgs,
-               trans_mats,
-               locations=None):
+               reg: Tensor,
+               points: Tensor,
+               labels: Tensor,
+               cam2imgs: Tensor,
+               trans_mats: Tensor,
+               locations: Tensor = None) -> Tuple[Tensor]:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/coders/smoke_bbox_coder.py:65]: locations: Optional[Tensor] = None

File: mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py [hunk 28]
```diff
@@ -29,13 +33,13 @@ class IoUNegPiecewiseSampler(RandomSampler):
...
     def __init__(self,
-                 num,
-                 pos_fraction=None,
-                 neg_piece_fractions=None,
-                 neg_iou_piece_thrs=None,
-                 neg_pos_ub=-1,
-                 add_gt_as_proposals=False,
-                 return_iou=False):
+                 num: int,
+                 pos_fraction: float = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py:37]: Optional[float] = None

File: mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py [hunk 29]
```diff
@@ -29,13 +33,13 @@ class IoUNegPiecewiseSampler(RandomSampler):
...
+                 neg_piece_fractions: list = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py:38]: Optional[list] = None

File: mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py [hunk 30]
```diff
@@ -29,13 +33,13 @@ class IoUNegPiecewiseSampler(RandomSampler):
...
+                 neg_iou_piece_thrs: list = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py:39]: Optional[list] = None

File: mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py [hunk 31]
```diff
@@ -127,11 +133,11 @@ def _sample_neg(self, assign_result, num_expected, **kwargs):
...
     def sample(self,
-               assign_result,
-               bboxes,
-               gt_bboxes,
-               gt_labels=None,
-               **kwargs):
+               assign_result: AssignResult,
+               bboxes: Tensor,
+               gt_bboxes: Tensor,
+               gt_labels: Tensor = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/samplers/iou_neg_piecewise_sampler.py:139]: Optional[Tensor]

File: mmdet3d/models/task_modules/voxel/voxel_generator.py [hunk 32]
```diff
@@ -36,33 +38,33 @@ def __init__(self,
...
 
     @property
-    def point_cloud_range(self):
+    def point_cloud_range(self) -> List[float]:
         """list[float]: Range of point cloud."""
         return self._point_cloud_range
 
     @property
-    def grid_size(self):
+    def grid_size(self) -> np.int64:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/task_modules/voxel/voxel_generator.py:63]: np.ndarray

**Comment by Xiangxu-0103** [comment]
Hi @A-new-b , please update this PR according to the comments and fix the lint issue.

**Review by Xiangxu-0103 — APPROVED** [review]

**Pull request merged**